src/backend/services/ExtractZipService.ts

```diff
@@ -0,0 +1,400 @@
+   */
+  get isCanceled(): boolean {
+    return this.#canceled
+  }
+
+  /**
+   * Get if is paused or not
+   * @returns {boolean} - Current state
+   */
+  get isPaused(): boolean {
+    return this.#paused
+  }
+
+  /**
+   * Gets the source ZIP file path.
+   * @returns {string} - The path to the ZIP file.
+   */
+  get source(): string {
+    return this.#zipFile
+  }
+
+  /**
+   * Gets the destination path where files will be extracted.
+   * @returns {string} - The destination path.
+   */
+  get destination(): string {
+    return this.#destinationPath
+  }
+
+  /**
+   * Check if can progress
+   * @returns {boolean}
+   */
+  get #canProgress(): boolean {
+    return !(this.#canceled || this.#paused)
+  }
+
+  get #isValid(): ExtractionValidation {
+    const isExisting = existsSync(this.#zipFile)
+
+    if (!isExisting) {
+      return ExtractionValidation.NOT_EXIST
+    }
+
+    if (/^[\\x00-\\x7F]+$/.test(this.#zipFile)) {
+      return ExtractionValidation.INVALID_FILE_NAME_ASCII
+    }
+
+    return ExtractionValidation.VALID
+  }
+
+  /**
+   * Pause the extraction process.
+   * @returns {void}
+   */
+  public pause(): void {
+    if (!this.isPaused) {
+      this.#paused = true
+
+      this.#readStream?.pause()
+
+      this.emit('paused', this.#computeProgress())
+    }
+  }
+
+  /**
+   * Resume the extraction process.
+   * @returns {Promise<boolean | void>}
+   */
+  public async resume(): Promise<boolean | void> {
+    if (!this.#extractionPromise) {
+      throw new Error('Extraction has not started or has already completed.')
+    }
+
+    if (this.isPaused) {
+      this.#paused = false
+
+      this.#readStream?.resume()
+    }
+
+    this.emit('resumed', this.#computeProgress())
+
+    return this.#extractionPromise
+  }
+
+  /**
+   * Cancels the extraction process.
+   * @returns {void}
+   */
+  public cancel() {
+    if (!this.#zipFileInstance?.isOpen) {
+      throw new Error('Extraction has not started or has already completed.')
+    }
+
+    this.#canceled = true
+
+    this.#readStream?.unpipe()
+
+    this.#readStream?.destroy(new Error('Extraction canceled'))
+
+    if (this.#zipFileInstance && this.#zipFileInstance.isOpen) {
+      this.#zipFileInstance.close()
+    }
+
+    this.emit('canceled')
+
+    rmSync(this.source, { recursive: true, force: true })
+
+    this.removeAllListeners()
+  }
+
+  /**
+   * Computes the progress of the extraction process.
+   * @private
+   * @returns {ExtractZipProgressResponse} - The progress details.
+   */
+  #computeProgress(): ExtractZipProgressResponse {
+    const progressPercentage = Math.min(
+      100,
+      (this.#processedSize / this.#totalSize) * 100
+    )
+    const elapsedTime = (Date.now() - this.#startTime) / 1000
+    const speed = this.#processedSize / elapsedTime
+
+    return {
+      progressPercentage,
+      speed,
+      totalSize: this.#totalSize,
+      processedSize: this.#processedSize
+    }
+  }
```

Comment:
might be good to round to nearest tenth of a percent here for `progressPercentage`. `elapsedTime` could be rounded to nearest second. not sure what units speed is but could round that too

Also we should rename these variables to include their units as it is only documented at the interface's comment and that should be documented on the variable name instead to keep the code readable

`elapsedTime` should be `elapsedTimeInSeconds`
`speed` should be `speedInMbPerSec` if those are the units
`totalSize` should be `totalSizeInBytes` or MB if those are the units. Same for `processedSize`

Reply:
Done